cmd/security-profiles-operator/main.go

```diff
@@ -102,31 +104,76 @@ func run(ctx *cli.Context) error {
 		return errors.Wrap(err, "get config")
 	}
 
-	ctrlOpts := ctrl.Options{
-		SyncPeriod: &sync,
+	fatalErrors := make(chan error)
+	done := make(chan bool)
+	sigHandler := ctrl.SetupSignalHandler()
+
+	// Daemonset managers
+	go func() {
+		ctrlOpts := ctrl.Options{
+			SyncPeriod: &sync,
+		}
+
+		if os.Getenv(restrictNSKey) != "" {
+			ctrlOpts.Namespace = os.Getenv(restrictNSKey)
+		}
+
+		mgr, err := ctrl.NewManager(cfg, ctrlOpts)
+		if err != nil {
+			fatalErrors <- errors.Wrap(err, "create manager")
+		}
+
+		if err := v1alpha1.AddToScheme(mgr.GetScheme()); err != nil {
+			fatalErrors <- errors.Wrap(err, "add core operator APIs to scheme")
+		}
+
+		if err := profile.Setup(ctx.Context, mgr, ctrl.Log.WithName("profile")); err != nil {
+			fatalErrors <- errors.Wrap(err, "setup profile controller")
+		}
+
+		setupLog.Info("starting manager")
+		if err := mgr.Start(sigHandler); err != nil {
+			fatalErrors <- errors.Wrap(err, "controller manager error")
+		}
+		close(done)
+	}()
+
+	// cluster managers (need leader election)
```

Comment:
The operator currently runs as a Daemonset and doesn't need leader election because it just runs the same operation on every node. If I'm getting this right, the new configmap controller spins up a separate pod to run the installer/uninstaller, and so it doesn't really need to be a Daemonset. I wonder if it would make sense to run the new controllers as a separate application, maybe in a Deployment? That would also allow us to split the RBAC and service accounts.

Reply:
That's actually how the original implementation of the selinux operator worked. It was a deployment that spawned the installer pods as a daemonset. This was just how I was able to adapt the code to this new code base. Ideally we would switch back to having the controllers on a deployment as that's what package managers such as the OLM (for Operator Hub) expect and require.